[Dx-962] Add FAQ For How To Distribute Analytics Keys Across Redis Clusters (#4018)
TykTechnologies/tyk-docs: tyk-docs/content/developer-support/frequently-asked-questions/how-to-reduce-cpu-usage-in-a-redis-cluster.md

---
title: "How To Reduce CPU Usage In A Redis Cluster"
date: 2024-01-22
tags: ["Analytics", "Distributed Analytics", "Redis", "Redis Shards", "analytics_config.enable_multiple_analytics_keys" ]
description: "Explains how to reduce CPU usage by configuring Tyk to use a Redis Cluster to distribute analytics keys to multiple Redis Shards"
---

> The illustration below highlights the scenario where a single Redis node is exhibiting high CPU usage of 1.20% within a Redis Cluster. One possible reason for this could be that analytics are enabled and the keys are being stored within a single Redis server.

Reviewer: ```suggestion
When you h
The illustration below highlights the scenario where a single Redis node is exhibiting high CPU usage of 1.20% within a Redis Cluster. One of the main possible reasons could be that Tyk Gateway is using only one node for storing the analytics. 

  analytics are enabled and the keys are being stored within a single Redis server.
```

> {{< img src="/img/faq/enable-multiple-analytics-keys/redis_single.png" >}}

Reviewer: too big

> Tyk distributes analytics to Redis storage. A high volume of analytics traffic can decrease performance, since all analytics keys are stored within one Redis server. 

Reviewer: `Tyk distributes analytics to Redis storage` - what does it mean? if it stores in one node, then what does it mean "Tyk distributes..."? do you mean store?

Author: Updated in section *## How does storing keys within a single Redis server contribute to high CPU usage?*

In Redis, *key sharding* is a term used to describe the practice of distributing data across multiple Redis instances or *shards* based on the keys. This feature is provided by [Redis Cluster](https://redis.io/docs/management/scaling/) and provides horizontal scalability and improved performance. 

Tyk supports configuring this behaviour so that analytics keys are distributed across multiple servers within a Redis cluster. The image below illustrates that CPU usage is reduced across two Redis servers after making this configuration change.

{{< img src="/img/faq/enable-multiple-analytics-keys/redis_distributed.png" >}}

> This FAQ explains how to configure Tyk to distribute analytic keys across multiple Redis instances.

Reviewer: "This FAQ explains how to configure" - not exactly. It also explains the reasons for the cpu load in only one node. Also, we can't write it half way through the page, doesn't make sense. 

Can we rephrase?

Author: Refactored introduction

---
title: "How To Reduce CPU Usage In A Redis Cluster"
date: 2024-01-22
tags: ["Analytics", "Distributed Analytics", "Redis", "Redis Shards", "analytics_config.enable_multiple_analytics_keys" ]
description: "Explains how to reduce CPU usage by configuring Tyk to use a Redis Cluster to distribute analytics keys to multiple Redis Shards"
---

This FAQ explains a potential reason for high CPU usage in a Redis node within a Redis Cluster and explains the steps that can be taken to resolve the issue.

## What does high CPU usage in a Redis node within a Redis Cluster mean?

> When a single Redis node within a Redis Cluster exhibits high CPU usage, it indicates that the CPU resources of that particular node are being heavily utilised compared to others in the cluster.

Reviewer: I think we should move the image redis_single.png to this section

> When a single Redis node within a Redis Cluster exhibits high CPU usage, it indicates that the CPU resources of that particular node are being heavily utilised compared to others in the cluster.

Reviewer: ```suggestion
When a single Redis node within a Redis Cluster exhibits high CPU usage, it indicates that the CPU resources of that particular node are being heavily utilised compared to others in the cluster.

The illustration below highlights the scenario where a single Redis node is exhibiting high CPU usage of 1.20% within a Redis Cluster. 

{{< img src="/img/faq/enable-multiple-analytics-keys/redis_single.png" width="768" alt="analytics keys stored in one Redis server" >}}
```

## What could be causing this high CPU usage?

One possible reason for high CPU usage in a single Redis node within a Redis Cluster is that analytics features are enabled and keys are being stored within that specific Redis node.

## How does storing keys within a single Redis server contribute to high CPU usage?

Storing keys within a single Redis server can lead to increased CPU usage because all operations related to those keys, such as retrieval, updates and analytics processing, are concentrated on that server. This can result in heavier computational loads on that particular node. This leads to high CPU usage.

## What can be done to address high CPU usage in this scenario?

Consider distributing the analytics keys across multiple Redis nodes within the cluster. This can help distribute the computational load more evenly, reducing the strain on any single node and potentially alleviating the high CPU usage.

> 

Reviewer: ```suggestion
```

> The illustration below highlights the scenario where a single Redis node is exhibiting high CPU usage of 1.20% within a Redis Cluster. 

Reviewer: ```suggestion
```

> {{< img src="/img/faq/enable-multiple-analytics-keys/redis_single.png" width="768" alt="analytics keys stored in one Redis server" >}}

Reviewer: ```suggestion
```

> Tyk stores analytics in Redis storage. A high volume of analytics traffic can decrease performance, since all analytics keys are stored within one Redis server. 

Reviewer: `Tyk stores analytics in Redis storage` - seems a bit too deep into the content, so say something so basic.

Author: updated
